### PR TITLE
perf: Improve ResourceDictionary performance

### PR DESCRIPTION
**Pull request by jeromelaban**
## PR Type

What kind of change does this PR introduce?
- Refactoring (no functional changes, no api changes)

## What is the new behavior?

- Improves the `ResourceDictionary` lookup performance by:
  - Caching key hash code, using pre-computed cache keys whenever possible during XAML materialization
  - Custom implementation of Dictionary based on .NET 6 `Dictionary<,>` to remove unused parts
  - Use of `(string, bool)` to discriminate types in the inner dictionary
  - Use of in parameters to avoid unnecessary copies
  - Depending on the platforms, the worst-case read improvements range from 30% to 300%.
- Restores Samples App logging for non-wasm targets
- Bumps AndroidX dependency versions to fix invalid VTable errors on recent VS builds

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/DatePickerIntegrationTests.cs [hunk 1]
```diff
@@ -469,6 +468,11 @@ public async Task CanOpenAndCloseUsingKeyboard()
 		[DataRow("ThaiCalendar")]
 		//[DataRow("UmAlQuraCalendar")]
 
+		// HebrewCalendar fails
+		// AssertFailedException: Expected string to be "11", but "13" differs near "3" (index 1).
+		// Expected string to be "February" with a length of 8, but "January" has a length of 7, differs near "Jan" (index 0).
+		//[DataRow("HebrewCalendar")]
+
```
Comment (kazo0) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/DatePickerIntegrationTests.cs:475]: FYI this should have hopefully been fixed by https://github.com/unoplatform/uno/pull/5439

**Review by kazo0 — APPROVED** [review]

**Review by ajpinedam — APPROVED** [review]

**Review by carldebilly — APPROVED** [review]
`goto:` instructions in your dictionary... that's unexpected.


:-)

File: src/SamplesApp/SamplesApp.Droid/SamplesApp.Droid.csproj [hunk 2]
```diff
@@ -102,7 +104,11 @@
     <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="1.1.1" />
     <PackageReference Include="Microsoft.Extensions.Logging.Filter" Version="1.1.1" />
     <PackageReference Include="Xamarin.AndroidX.Lifecycle.LiveData" Version="2.2.0.2" />
-    <PackageReference Include="Xamarin.AndroidX.Fragment" Version="1.1.0.1" />
+    <PackageReference Include="Xamarin.AndroidX.Fragment" Version="1.3.0.1" />
+    <PackageReference Include="Uno.BenchmarkDotNet" Version="0.11.6-develop" />
```
Comment (davidjohnoliver) [inline src/SamplesApp/SamplesApp.Droid/SamplesApp.Droid.csproj:108]: Very nice!

File: src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/DependenciesInitializer.cs [hunk 3]
```diff
@@ -16,24 +16,36 @@ namespace Uno.UI.SourceGenerators
 {
 	internal class DependenciesInitializer
 	{
+		private static object _gate = new object();
+
 		internal static void Init()
```
Comment (davidjohnoliver) [inline src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/DependenciesInitializer.cs:21]: Is there a check elsewhere that this only runs once?

File: src/Uno.UI.BindingHelper.Android/Uno.UI.BindingHelper.Android.csproj [hunk 4]
```diff
@@ -53,7 +54,14 @@
...
 
-	<Target Name="_CompileUnoJavaCreateOutputs" BeforeTargets="Build" AfterTargets="Restore">
+	<!--
+		The dependency on _CleanJavaGenerated is required to avoid partial assembly
+		rebuild that create "missing UnoViewGroup" errors. This situation can happen
+		when modifying an external msbuild props/targets file, such as Directory.Build.props
+		then building the project. Cleaning the generated files ensures that the whole
+		binding assembly is generated properly.
+	-->
+	<Target Name="_CompileUnoJavaCreateOutputs" BeforeTargets="Build" AfterTargets="Restore" DependsOnTargets="_CleanJavaGenerated">
```
Comment (davidjohnoliver) [inline src/Uno.UI.BindingHelper.Android/Uno.UI.BindingHelper.Android.csproj:64]: 🎉

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 5]
```diff
@@ -282,25 +354,29 @@ private bool ContainsKeyThemeMerged(object key)
 		private void CopyFrom(ResourceDictionary source)
 		{
 			_values.Clear();
-			MergedDictionaries.Clear();
+			_mergedDictionaries.Clear();
 			_themeDictionaries?.Clear();
 
-			_values.AddRange(source);
-			MergedDictionaries.AddRange(source.MergedDictionaries);
+			_values.AddRange(source._values);
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:360]: am I reading this correctly: with this change it won't eagerly materialize on copy?

File: src/Uno.UI/UI/Xaml/SpecializedResourceDictionary.cs [hunk 6]
```diff
@@ -0,0 +1,1404 @@
...
+				}
+				else if (key is Type t)
+				{
+					Key = t.ToString();
+					IsType = true;
+					HashCode = (uint)(t.GetHashCode() ^ IsType.GetHashCode());
+				}
+				else
+				{
+					Key = key.ToString();
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/SpecializedResourceDictionary.cs:63]: I feel like there could be the potential for duplicate keys in corner cases involving non-string/non-type objects used as keys (ie two different objects returning the same `ToString()` value). (Disclaimer: I don't even know if UWP properly supports non-string/non-type objects as ResourceDictionary keys.)

File: src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/DependenciesInitializer.cs (hunk 3 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/DependenciesInitializer.cs:21]: It does not really need to be running once, additional runs will only return the same assemblies.

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs (hunk 5 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:360]: No, this change does not impact this behavior. The new internal dictionary does not expose AddRange the same way the previous one did, hence the change. Non-materializing copy is for another PR.

**Pull request merged**